Description: Retain support for cloud-init -f FILE argument on the commandline
 Upstream has disallowed the use of -f/--file argument as a generic argument on
 the command line before the positional cloud-init subcommand is provided.
 Retain the ability to provide -f/--file on the commandline before the
 subcommand init, modules or single.
Author: Chad Smith <chad.smith@canonical.com>
Origin: backport
Bug: https://bugs.launchpad.net/ubuntu/+source/cloud-init/+bug/2064300
Last-Update: 2024-04-30 <YYYY-MM-DD, last update of the meta-information, optional>
--- a/cloudinit/cmd/main.py
+++ b/cloudinit/cmd/main.py
@@ -138,6 +138,11 @@ def extract_fns(args):
     # since it would of broke if it couldn't have
     # read that file already...
     fn_cfgs = []
+    if args.main_files:
+        # Append common --file argument to beginning of list when present.
+        # Avoid namespace collision with same --file arg defined in subcommands
+        for fh in args.main_files:
+            fn_cfgs.append(os.path.realpath(fh.name))
     if args.files:
         for fh in args.files:
             # The realpath is more useful in logging
@@ -1055,6 +1060,17 @@ def main(sysv_args=None):
         default=False,
     )
     parser.add_argument(
+        "--file",
+        "-f",
+        action="append",
+        dest="main_files",
+        help=(
+            "Use additional yaml configuration files for subcommands: init,"
+            " modules or single"
+        ),
+        type=argparse.FileType("rb"),
+    )
+    parser.add_argument(
         "--force",
         action="store_true",
         help=(
--- a/tests/unittests/cmd/test_main.py
+++ b/tests/unittests/cmd/test_main.py
@@ -14,7 +14,8 @@ from cloudinit.cmd import main
 from cloudinit.util import ensure_dir, load_text_file, write_file
 
 MyArgs = namedtuple(
-    "MyArgs", "debug files force local reporter subcommand skip_log_setup"
+    "MyArgs",
+    "debug main_files files force local reporter subcommand skip_log_setup",
 )
 
 
@@ -108,6 +109,7 @@ class TestMain:
             files = None
         cmdargs = MyArgs(
             debug=False,
+            main_files=None,
             files=files,
             force=False,
             local=False,
@@ -144,6 +146,7 @@ class TestMain:
         write_file(cloud_cfg_file, safeyaml.dumps(cfg))
         cmdargs = MyArgs(
             debug=False,
+            main_files=None,
             files=None,
             force=False,
             local=False,
@@ -340,6 +343,7 @@ class TestMain:
         write_file(cloud_cfg_file, safeyaml.dumps(cfg))
         cmdargs = MyArgs(
             debug=False,
+            main_files=None,
             files=None,
             force=False,
             local=False,
